compiler/src/iree/compiler/Codegen/SPIRV/SPIRVLinkExecutables.cpp

```diff
@@ -7,21 +7,41 @@
 #include "iree/compiler/Codegen/SPIRV/PassDetail.h"
 #include "iree/compiler/Codegen/SPIRV/Passes.h"
 #include "iree/compiler/Codegen/Utils/LinkingUtils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALTypes.h"
 #include "iree/compiler/Utils/ModuleUtils.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Debug.h"
 #include "llvm/Support/FormatVariadic.h"
 #include "mlir/Dialect/SPIRV/IR/SPIRVOps.h"
 #include "mlir/Pass/Pass.h"
 
+#define DEBUG_TYPE "iree-spirv-link-executable"
+
 namespace mlir::iree_compiler {
 
+namespace IREE::HAL {
+// MLIR attributes are pointers to uniqued data. So we can just compare the
+// pointer for order.
+bool operator<(const ExecutableTargetAttr &a, const ExecutableTargetAttr &b) {
+  return a.getAsOpaquePointer() < b.getAsOpaquePointer();
+}
+} // namespace IREE::HAL
+
 namespace {
 
+using IREE::HAL::ExecutableTargetAttr;
+
+bool isSPIRVBasedBackend(StringRef backend) {
+  return backend.starts_with("vulkan") || backend.starts_with("metal") ||
```

Comment:
is there something else you can key off instead? like the presence of spv attributes on the module?
(lists like this prevent out-of-tree targets from working)

Reply:
Good point about out-of-tree targets. I switched to check the existance of `spirv.target_env`, which should be more permissive there.